### ISSUE TITLE
never_loop declares that a flattened loop never actually runs when it does

### ISSUE THREAD

**Issue by zkldi** [rust-lang/rust-clippy#11774]
### Summary

I had a helpful clippy lint: `clippy::manual_flatten` that I applied, but after changing my iterator to a flattened one I got a `never_loop` error.

This is a false positive as the loop most definitely still runs.

### Lint Name

never_loop

### Reproducer

I tried this code:

```rust
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Test(pub [Option<i32>; 4]);

impl Test {
	pub fn needs_flatten(&self) -> i32 {
		for e in &self.0 {
			// warning: unecessary if let ... clippy::manual_flatten
			if let Some(e) = e {
				return *e;
			}
		}

		panic!("heck")
	}
	
	pub fn flattened(&self) -> i32 {
		// error: this loop never actually loops ... clippy::never_loop
		for e in self.0.iter().flatten() {
			return *e;
		}

		panic!("heck")
	}
}

fn main() {
    let t = Test([None, Some(2), None, Some(4)]);
    
	// both calls output `2`.
    dbg!(t.needs_flatten(), t.clippy_wtf());
}
```

I saw this happen:

```
error: this loop never actually loops
  --> src/main.rs:16:3
   |
16 | /         for e in self.0.iter().flatten() {
17 | |             return *e;
18 | |         }
   | |_________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#never_loop
   = note: `#[deny(clippy::never_loop)]` on by default
help: if you need the first element of the iterator, try writing
   |
16 |         if let Some(e) = self.0.iter().flatten().next() {
   |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I expected to see this happen:

no error, this loop definitely runs

### Version

```text
rustc 1.74.0-nightly (62ebe3a2b 2023-09-08)
binary: rustc
commit-hash: 62ebe3a2b177d50ec664798d731b8a8d1a9120d1
commit-date: 2023-09-08
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

**Comment by Jarcho** [comment]
The second version doesn't loop; the body runs at most once. The whole function is equivalent to `self.0.iter().flatten().next().expect("heck")`.

**Comment by zkldi** [comment]
Oh that's complete silliness on my part!

I misread never loop as "this block never runs", but it actually means "this block never loops"

Total mistake on my part and clippy is right here.
